roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
@stevekuznetsov,  thanks for the explanations on how `patch` works.  I researched early on in the 3.1-3.2 days and decided with its behavior it wasn't going to meet our needs.

When lib_openshift makes the changes it does use the resource version of an object.  We have run into issues when these have been modified and the module will fail when openshift refuses to replace with an old resource version.  This is undesirable but is something we will have to handle.  Currently we do not handle this and force the user to rerun to apply state.  This is not a good user experience but it is safer to guarantee state.  I imagine we will want to build in retry loops and timeouts.  This is the first version of the modules and they will continue to develop.

The questions about `patch` still remain.
Can `patch` perform multiple edits at a single time?  If not, are you proposing we run multiple `patch` commands to achieve state?  Isn't this susceptible to the race conditions you mentioned above?
How does `patch` handle race conditions?  It seems to me that the situation you described would result in whoever made the edits last would win.  This is kind of the same behavior as `replace` except we attempt to obey resource versions.  I'm not sure we can guarantee state if multiple administrators are attempting to modify the same object if its not done with a transaction.  As soon as `patch` makes changes and another call to `patch` occurs you run the risk of someone making a different change and now the module cannot, with confidence, return that the desired "state" was achieved.

Good discussion going here.